lib/mayaUsd/render/vp2RenderDelegate/material.cpp

```diff
@@ -796,6 +699,128 @@ change tracker for use in the first sync of this prim.
 */
 HdDirtyBits HdVP2Material::GetInitialDirtyBitsMask() const { return HdMaterial::AllDirty; }
 
+/*! \brief  Applies VP2-specific fixes to the material network.
+ */
+void HdVP2Material::_ApplyVP2Fixes(HdMaterialNetwork& outNet, const HdMaterialNetwork& inNet)
+{
+    // To avoid relocation, reserve enough space for possible maximal size. The
+    // output network is temporary C++ object that will be released after use.
+    const size_t numNodes = inNet.nodes.size();
+    const size_t numRelationships = inNet.relationships.size();
+
+    size_t nodeCounter = 0;
+
+    _nodePathMap.clear();
+    _nodePathMap.reserve(numNodes);
+
+    HdMaterialNetwork tmpNet;
+    tmpNet.nodes.reserve(numNodes);
+    tmpNet.relationships.reserve(numRelationships);
+
+    // Replace the authored node paths with simplified paths in the form of "node#". By doing so
+    // we will be able to reuse shader effects among material networks which have the same node
+    // identifiers and relationships but different node paths, reduce shader compilation overhead
+    // and enable material consolidation for faster rendering.
+    for (const HdMaterialNode& node : inNet.nodes) {
+        tmpNet.nodes.push_back(node);
+
+        HdMaterialNode& outNode = tmpNet.nodes.back();
+        outNode.path = SdfPath(kNodePathPrefix + std::to_string(++nodeCounter));
```

Comment:
That is a very good idea as it will collapse networks with similar topologies and is really safe to use in the case of UsdPreviewSurface since all nodes have dissimilar parameter names and will probably never result in ambiguity.

However, for MaterialX, where we can have many math nodes where all the inputs are called "input1" and " input2" and all outputs called "out" you will merge together networks that differ only by the use of an "add" instead of a "multiply" node. The fix is actually simple. Instead of using a single "node" string + number, you should be using info:id + number.

Reply:
Good point. I was actually doing the approach with identifier + number and thought it was not necessary for UsdShade material. The change is very small, so we can do it when we implement MaterialX.

Reply:
Hmm... Second thought. I can do it now because it is actually also safer than the "node" prefix.

Reply:
Yes, we should not defer this work.

Reply:
The follow-up of this conversation is at https://github.com/Autodesk/maya-usd/pull/1048 .